nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/SubtypesJsonEnricher.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * @since 11.5
+ */
+@Setup(mode = SINGLETON, priority = REFERENCE)
+public class SubtypesJsonEnricher extends AbstractJsonEnricher<DocumentModel> {
+
+    public static final String NAME = "subtypes";
+
+    public SubtypesJsonEnricher() {
+        super(NAME);
+    }
+
+    @Override
+    public void write(JsonGenerator jg, DocumentModel enriched) throws IOException {
+        SchemaManager schemaManager = Framework.getService(SchemaManager.class);
+        Collection<String> subtypes = computeSubtypes(enriched);
+        jg.writeFieldName(NAME);
+        jg.writeStartArray();
+        for (String subtype : subtypes) {
+            jg.writeStartObject();
+            jg.writeStringField("type", subtype);
+            jg.writeArrayFieldStart("facets");
+            for (String facet : schemaManager.getDocumentType(subtype).getFacets()) {
+                jg.writeString(facet);
+            }
+            jg.writeEndArray();
+            jg.writeEndObject();
+        }
+        jg.writeEndArray();
+    }
+
+    protected Collection<String> computeSubtypes(DocumentModel enriched) {
+        Collection<String> defaultSubtypes = enriched.getDocumentType().getAllowedSubtypes();
+        if (enriched.hasFacet(UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_FACET)) {
+            return computeLocalConfigurationSubtypes(enriched, defaultSubtypes);
+        }
+        return defaultSubtypes;
+    }
+
+    protected Collection<String> computeLocalConfigurationSubtypes(DocumentModel enriched,
+            Collection<String> defaultSubtypes) {
+        Collection<String> subtypes = new HashSet<String>();
+        Boolean denyAllTypes = (Boolean) enriched.getPropertyValue(
+                UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_DENY_ALL_TYPES_PROPERTY);
+        if (BooleanUtils.isNotTrue(denyAllTypes)) {
+            String[] allowedTypesProperty = (String[]) enriched.getPropertyValue(
+                    UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_ALLOWED_TYPES_PROPERTY);
+            String[] deniedTypesProperty = (String[]) enriched.getPropertyValue(
+                    UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_DENIED_TYPES_PROPERTY);
+            List<String> allowedTypes = new ArrayList<>();
+            List<String> deniedTypes = new ArrayList<>();
+            if (allowedTypesProperty != null) {
+                allowedTypes = Arrays.asList(allowedTypesProperty);
+            }
+            if (deniedTypesProperty != null) {
+                deniedTypes = Arrays.asList(deniedTypesProperty);
+            }
+            if (!(allowedTypes.isEmpty() && deniedTypes.isEmpty())) {
+                for (String subtype : defaultSubtypes) {
+                    if (!deniedTypes.contains(subtype) && (allowedTypes.isEmpty() || allowedTypes.contains(subtype))) {
+                        subtypes.add(subtype);
+                    }
+                }
+            }
```

Comment:
`allowedTypes.isEmpty()` in the second if is always false, you can remove it.

We could use the Stream API there, with something such as below:
```java
import static java.util.stream.Collectors.toSet;

return defaultSubtypes.stream().filter(s -> !deniedTypes.contains(s)).filter(allowedTypes::contains).collect(toSet());
```

The proposed return will make the `subtypes` useless, you can replace the last return with a `Collections.emptySet()`.

Reply:
`allowedTypes.isEmpty()` could be true if we only put `deniedTypes` so this case has to be considered. In the case where `allowedTypes` is empty we need to consider all the subtypes defined by the document type.